Selenium tests

un-tone opened the pull request:
For https://github.com/StarcounterApps/Website/issues/60.

The tests cover the following issues: https://github.com/Starcounter/level1/issues/4115, https://github.com/StarcounterApps/Website/issues/44.

I will add new tests later for the https://github.com/StarcounterApps/Website/issues/58.

warpech commented on src/WebsiteProvider.Tests/packages.config:
```diff
@@ -0,0 +1,6 @@
+  <package id="NUnit" version="3.6.1" targetFramework="net45" />
+  <package id="Selenium.Support" version="3.3.0" targetFramework="net45" />
+  <package id="Selenium.WebDriver" version="3.3.0" targetFramework="net45" />
+</packages>
```
Please add NuGet package resolving to `build.bat`

un-tone replied:
Added.

warpech commented on Website.sln:
```diff
@@ -20,22 +20,61 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Website", "src\Website\Webs
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebsiteProvider", "src\WebsiteProvider\WebsiteProvider.csproj", "{050D34C2-F3ED-4712-8D9E-A47160B41DBB}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WebsiteProvider.Tests", "src\WebsiteProvider.Tests\WebsiteProvider.Tests.csproj", "{8EDF4429-251A-416D-BB68-93F227191BCF}"
```
Please add the test projects and helper apps in `test` directory, not `src`. That is how KitchenSink and Launcher have it. It will allow us to use the same testing scripts.

un-tone replied:
Moved test projects to `test` directory

un-tone commented:
I added `test.bat` script and testing instructions.

> For the start two comments regarding the project structure and batch files.

Fixed too.

> We will need to add more tests, though.

Sure.

warpech commented:
Something wrong with `sln` or `csproj`. Trying to build and run on a freshly cloned repo. I am getting a warning:

> The project "Website" is not selected for building in solution configuration "Debug|Any CPU".
> The project "WebsiteProvider" is not selected for building in solution configuration "Debug|Any CPU".

Full output:

```txt
c:\www\Website>build.bat
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0.26228.9
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
MSBuild auto-detection: using msbuild version '4.0' from 'C:\Windows\Microsoft.NET\Framework\v4.0.30319'.
Restoring NuGet package NUnit.ConsoleRunner.3.6.1.
Restoring NuGet package NUnit.3.6.1.
Restoring NuGet package Selenium.Support.3.3.0.
Restoring NuGet package Selenium.WebDriver.3.3.0.
Adding package 'NUnit.ConsoleRunner.3.6.1' to folder 'c:\www\Website\packages'
Adding package 'Selenium.Support.3.3.0' to folder 'c:\www\Website\packages'
Adding package 'NUnit.3.6.1' to folder 'c:\www\Website\packages'
Added package 'Selenium.Support.3.3.0' to folder 'c:\www\Website\packages'
Adding package 'Selenium.WebDriver.3.3.0' to folder 'c:\www\Website\packages'
Added package 'NUnit.ConsoleRunner.3.6.1' to folder 'c:\www\Website\packages'
Added package 'Selenium.WebDriver.3.3.0' to folder 'c:\www\Website\packages'
Added package 'NUnit.3.6.1' to folder 'c:\www\Website\packages'

NuGet Config files used:
    C:\Users\warpe\AppData\Roaming\NuGet\NuGet.Config

Feeds used:
    C:\Users\warpe\AppData\Local\NuGet\Cache
    C:\Users\warpe\.nuget\packages\
    https://api.nuget.org/v3/index.json

Installed:
    4 package(s) to packages.config projects
Microsoft (R) Build Engine version 15.1.548.43366
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
Build started 4/6/2017 9:15:20 PM.
Project "c:\www\Website\Website.sln" on node 1 (default targets).
ValidateSolutionConfiguration:
  Building solution configuration "Debug|Any CPU".
ValidateProjects:
  The project "Website" is not selected for building in solution configuration "Debug|Any CPU".
  The project "WebsiteProvider" is not selected for building in solution configuration "Debug|Any CPU".
  The project "WebsiteProvider_AcceptanceHelperOne" is not selected for building in solution configuration "Debug|Any C
  PU".
  The project "WebsiteProvider_AcceptanceHelperTwo" is not selected for building in solution configuration "Debug|Any C
  PU".
Project "c:\www\Website\Website.sln" (1) is building "c:\www\Website\test\WebsiteProvider.Tests\WebsiteProvider.Tests.c
sproj" (2) on node 1 (default targets).
PrepareForBuild:
  Creating directory "bin\Debug\".
  Creating directory "obj\Debug\".
GenerateTargetFrameworkMonikerAttribute:
Skipping target "GenerateTargetFrameworkMonikerAttribute" because all output files are up-to-date with respect to the i
nput files.
CoreCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\bin\Roslyn\csc.exe /noconfig /nowarn:1701,
  1702 /nostdlib+ /errorreport:prompt /warn:4 /define:DEBUG;TRACE /highentropyva+ /reference:"C:\Program Files (x86)\Re
  ference Assemblies\Microsoft\Framework\.NETFramework\v4.5\mscorlib.dll" /reference:c:\www\Website\packages\NUnit.3.6.
  1\lib\net45\nunit.framework.dll /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFrame
  work\v4.5\System.Core.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\
  v4.5\System.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\Syste
  m.Drawing.dll" /reference:c:\www\Website\packages\Selenium.WebDriver.3.3.0\lib\net40\WebDriver.dll /reference:c:\www\
  Website\packages\Selenium.Support.3.3.0\lib\net40\WebDriver.Support.dll /debug+ /debug:full /filealign:512 /optimize-
   /out:obj\Debug\WebsiteProvider.Tests.dll /ruleset:"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Tea
  m Tools\Static Analysis Tools\\Rule Sets\MinimumRecommendedRules.ruleset" /subsystemversion:6.00 /target:library /utf
  8output Properties\AssemblyInfo.cs Test\BaseTest.cs Test\PageLoadTests.cs Ui\AcceptanceHelperTwoMasterPage.cs Ui\Acce
  ptanceHelperOneMasterPage.cs Ui\BasePage.cs Utilities\Config.cs Utilities\Screenshot.cs Utilities\WebDriverManager.cs
   Utilities\WebDriverEventListener.cs "C:\Users\warpe\AppData\Local\Temp\.NETFramework,Version=v4.5.AssemblyAttributes
  .cs"
  Using shared compilation with compiler from directory: C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\
  MSBuild\15.0\bin\Roslyn
_CopyFilesMarkedCopyLocal:
  Copying file from "c:\www\Website\packages\NUnit.3.6.1\lib\net45\nunit.framework.dll" to "bin\Debug\nunit.framework.d
  ll".
  Copying file from "c:\www\Website\packages\Selenium.WebDriver.3.3.0\lib\net40\WebDriver.dll" to "bin\Debug\WebDriver.
  dll".
  Copying file from "c:\www\Website\packages\Selenium.Support.3.3.0\lib\net40\WebDriver.Support.dll" to "bin\Debug\WebD
  river.Support.dll".
  Copying file from "c:\www\Website\packages\NUnit.3.6.1\lib\net45\nunit.framework.xml" to "bin\Debug\nunit.framework.x
  ml".
  Copying file from "c:\www\Website\packages\Selenium.WebDriver.3.3.0\lib\net40\WebDriver.xml" to "bin\Debug\WebDriver.
  xml".
  Copying file from "c:\www\Website\packages\Selenium.Support.3.3.0\lib\net40\WebDriver.Support.xml" to "bin\Debug\WebD
  river.Support.xml".
CopyFilesToOutputDirectory:
  Copying file from "obj\Debug\WebsiteProvider.Tests.dll" to "bin\Debug\WebsiteProvider.Tests.dll".
  WebsiteProvider.Tests -> c:\www\Website\test\WebsiteProvider.Tests\bin\Debug\WebsiteProvider.Tests.dll
  Copying file from "obj\Debug\WebsiteProvider.Tests.pdb" to "bin\Debug\WebsiteProvider.Tests.pdb".
Done Building Project "c:\www\Website\test\WebsiteProvider.Tests\WebsiteProvider.Tests.csproj" (default targets).

Done Building Project "c:\www\Website\Website.sln" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.92

c:\www\Website>run.bat

ScErrExecutableNotFound (SCERR10019): The executable file cound not be found. File: "c:\www\Website\src\Website\bin\Debug\Website.exe"
Version: 2.3.0.6120.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10019.

Type "star -h" to see help
Type "star 10019" to launch the help page for error 10019

ScErrExecutableNotFound (SCERR10019): The executable file cound not be found. File: "c:\www\Website\src\WebsiteProvider\bin\Debug\WebsiteProvider.exe"
Version: 2.3.0.6120.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10019.

Type "star -h" to see help
Type "star 10019" to launch the help page for error 10019

c:\www\Website>
```

un-tone commented:
> Something wrong with `sln` or `csproj`

Fixed.

warpech commented:
Working good now! 

@un-tone What do you recommend, should we merge it now and make sure that next fixes always include tests?

warpech commented on README.md:
```diff
@@ -108,6 +108,32 @@ This is how it looks as of Jan 2017:
+1. Start Selenium Remote Driver: `java -jar selenium-server-standalone-3.*.jar`
+2. Open `Website.sln` in Visual Studio and enable Test Explorer (Test > Window > Test Explorer)
+3. You need to install NUnit 3 Test Adapter in VS addon window in order to see tests in Test Explorer window
+3. Start the KitchenSink app
```
Change to the correct app names

un-tone replied:
Oh, thanks. Fixed.

warpech commented on test.bat:
```diff
@@ -0,0 +1,33 @@
+
+:: Start the test
+IF NOT EXIST "%~dp0packages\NUnit.ConsoleRunner.3.6.1\" (ECHO Error: Cannot find NUnit Console Runner. Build the project to restore the packages && PAUSE && EXIT /B 1)
+%~dp0packages\NUnit.ConsoleRunner.3.6.1\tools\nunit3-console.exe %~dp0test\WebsiteProvider.Tests\bin\Debug\WebsiteProvider.Tests.dll --noheader --params Browsers=Chrome
```
Why testing only in Chrome? Should we add Firefox here as well?

un-tone replied:
Added.

un-tone commented:
> should we merge it now and make sure that next fixes always include tests
Yes, I prefer this way.

warpech commented on test.bat:
```diff
@@ -25,7 +25,7 @@ IF ERRORLEVEL 1 EXIT /B 1
 :: Start the test
 IF NOT EXIST "%~dp0packages\NUnit.ConsoleRunner.3.6.1\" (ECHO Error: Cannot find NUnit Console Runner. Build the project to restore the packages && PAUSE && EXIT /B 1)
-%~dp0packages\NUnit.ConsoleRunner.3.6.1\tools\nunit3-console.exe %~dp0test\WebsiteProvider.Tests\bin\Debug\WebsiteProvider.Tests.dll --noheader --params Browsers=Chrome
+%~dp0packages\NUnit.ConsoleRunner.3.6.1\tools\nunit3-console.exe %~dp0test\WebsiteProvider.Tests\bin\Debug\WebsiteProvider.Tests.dll --noheader --params Browsers=Chrome,Firefox
```
@KmaTTi89 [says](https://github.com/StarcounterApps/KitchenSink/issues/116#issuecomment-292474818) that we will have the same effect by just removing `--params Browsers=Chrome,Firefox` completely.

un-tone replied:
Fixed, thanks
